src/authentication/authentication-hapi-plugin.ts

```diff
@@ -38,23 +47,174 @@ class AuthenticationHapiPlugin extends HapiPlugin {
-    const teams = getUserTeams(user.id, this.gitlab);
-    return reply(teams);
+    try {
+      const credentials = request.auth.credentials as AccessToken;
+      // The email has been validated in validateUser at this point
+      const user = await this.gitlab.getUserByEmailOrUsername(credentials.email!);
+      const teams = await this.gitlab.getUserGroups(user.id);
+      return reply(teams[0]); // NOTE: we only support a single team for now
+    } catch (error) {
+      this.logger.error(`Can't fetch user or team`, error);
+      return reply(Boom.wrap(error, 404));
+    }
+  }
+
+  public async getTeamTokenHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    try {
+      const teamId = parseInt(request.paramsArray[0], 10);
+      const teamToken: TeamToken[] = await teamTokenQuery(this.db, undefined, teamId);
+      if (!teamToken.length) {
+        throw new Error(`No token found for team ${teamId}`);
+      }
+      return reply(teamToken[0]);
+    } catch (error) {
+      return reply(Boom.notFound(error.message));
+    }
+  }
+
+  public async createTeamTokenHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    try {
+      const teamId = parseInt(request.paramsArray[0], 10);
+      const teamToken = await generateAndSaveTeamToken(teamId, this.db);
+      return reply(teamToken).code(201);
+    } catch (error) {
+      return reply(Boom.badRequest(error.message));
+    }
+  }
+
+  public async signupHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    let email: string | undefined;
+    let credentials: AccessToken | undefined;
+    try {
+      credentials = request.auth.credentials as AccessToken;
+      email = credentials.email;
+      if (!validateEmail(email)) {
+        // Fall back to fetching the email from Auth0
+        email = await this.tryGetEmailFromAuth0((request.auth as any).token);
+      }
+      if (!validateEmail(email)) {
+        throw new Error(`Invalid email ${email}`);
+      }
+      const teamToken = credentials[teamTokenClaimKey]!;
+      const teamId = await getTeamIdWithToken(teamToken, this.db);
+      const team = await this.gitlab.getGroup(teamId);
+      const password = generatePassword();
+      const {id, idp} = parseSub(credentials.sub);
+      const user = await this.gitlab.createUser(
+        email,
+        password,
+        credentials.sub,
+        email,
+        id,
+        idp,
+      );
+      await this.gitlab.addUserToGroup(user.id, teamId);
+      return reply({
+        team,
+        password,
+      }).code(201); // created
+    } catch (error) {
+      const message = `Problems on signup for user ${email}`;
+      this.logger.error(message, {
+        error,
+        credentials,
+      });
+      return reply(Boom.badRequest(message));
+    }
   }
 
   public async validateUser(
     payload: AccessToken,
     _request: any,
     callback: (err: any, valid: boolean, credentials?: any) => void,
   ) {
```

Comment:
There are no actual checks performed for the `sub` claim before this point, it might be missing or of invalid format.